File: modules/admin-ui/src/main/resources/web/admin/common/js/content/page/region/TextComponent.ts
module api.content.page.region {

    export class TextComponent extends Component implements api.Equitable, api.Cloneable {

        private text: string;

        public static PROPERTY_TEXT: string = "text";

        constructor(builder?: TextComponentBuilder) {
            super(builder);
            if (builder) {
                this.setText(builder.text, true);
            }
        }

        getText(): string {
            return this.text;
        }

        setText(value?: string, silent?: boolean) {
            this.text = api.util.StringHelper.isBlank(value) ? undefined : value;

            if (!silent) {
                this.notifyPropertyChanged(TextComponent.PROPERTY_TEXT);
            }
        }

        doReset() {
            this.setText();
        }

        isEmpty(): boolean {
            return api.util.StringHelper.isBlank(this.text);
        }

        toJson(): ComponentTypeWrapperJson {

            var json: TextComponentJson = <TextComponentJson>super.toComponentJson();
            json.text = this.text != null ? this.text : null;

            return <ComponentTypeWrapperJson> {
                TextComponent: json
            };
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, TextComponent)) {
                return false;
            }

            var other = <TextComponent>o;

            if (!super.equals(o)) {
                return false;
            }

            if (!api.ObjectHelper.stringEquals(this.text, other.text)) {
                return false;
            }

            return true;
        }

        clone(): TextComponent {
            return new TextComponentBuilder(this).build();
        }
    }

    export class TextComponentBuilder extends ComponentBuilder<TextComponent> {

        text: string;

        constructor(source?: TextComponent) {

            super(source);

            if (source) {
                this.text = source.getText();
            }
        }

        public fromJson(json: TextComponentJson, region: Region): TextComponentBuilder {

            if (json.text) {
                this.setText(json.text);
            }

            this.setName(json.name ? new ComponentName(json.name) : null);
            this.setParent(region);

            return this;
        }

        public setText(value: string): TextComponentBuilder {
            this.text = value;
            return this;
        }

        public build(): TextComponent {
            return new TextComponent(this);
        }
    }
}

File: modules/core-api/src/main/java/com/enonic/xp/content/page/region/TextComponent.java
package com.enonic.xp.content.page.region;


import java.util.Objects;

public class TextComponent
    extends Component
{
    private String text;

    protected TextComponent( final Builder builder )
    {
        super( builder );
        this.text = builder.text;
    }

    public static Builder newTextComponent()
    {
        return new Builder();
    }

    public static Builder newTextComponent( final TextComponent source )
    {
        return new Builder( source );
    }

    @Override
    public Component copy()
    {
        return newTextComponent( this ).build();
    }

    @Override
    public ComponentType getType()
    {
        return TextComponentType.INSTANCE;
    }

    public String getText()
    {
        return text;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }
        if ( !super.equals( o ) )
        {
            return false;
        }

        final TextComponent that = (TextComponent) o;

        if ( text != null ? !text.equals( that.text ) : that.text != null )
        {
            return false;
        }

        return true;
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( super.hashCode(), text );
    }

    public static class Builder
        extends Component.Builder

    {
        private String text;

        Builder()
        {
            // Default
        }

        private Builder( final TextComponent source )
        {
            super( source );
            text = source.text;
        }

        @Override
        public Builder name( ComponentName value )
        {
            this.name = value;
            return this;
        }

        public Builder name( String value )
        {
            this.name = value != null ? new ComponentName( value ) : null;
            return this;
        }

        public Builder text( String value )
        {
            this.text = value;
            return this;
        }

        public TextComponent build()
        {
            return new TextComponent( this );
        }
    }
}
